src/crc.rs

```diff
@@ -45,26 +43,18 @@ pub(crate) fn valid_checksum(data: &[u8], expected: u16) -> bool {
 
 pub(crate) fn push_crc(data: &mut Vec<u8>) {
     let crc = crc16(data);
-    data.write_u16::<LittleEndian>(crc).unwrap();
+    data.extend(u16::to_le_bytes(crc));
 }
 
 pub(crate) fn extract_crc(data: &mut Vec<u8>) -> Result<u16> {
-    let crc_end_idx = if data.len() >= 2 {
-        data.len() - 2
-    } else {
+    if data.len() < 2 {
         return Err(Error::new(
             ErrorKind::ChecksumFailure,
             Some("CRC data vector contains less than two characters"),
         ));
-    };
-    let crc_bytes = data[..crc_end_idx].to_vec();
+    }
 
-    let mut reader = Cursor::new(crc_bytes);
-    (*data).truncate(crc_end_idx);
-    reader.read_u16::<LittleEndian>().map_err(|e| {
-        Error::new(
-            ErrorKind::ChecksumFailure,
-            Some(format!("failed to read u16: {e}").as_str()),
-        )
-    })
+    let crc = u16::from_le_bytes(data[..2].try_into().unwrap());
+    data.truncate(data.len() - 2);
```

Comment:
I'm confident that `data.len()` isn't going to underflow here, but I would prefer that the snippet with `data[..2].try_into().unwrap()` returned an error instead of panicking. You might be able to just do `data[..2].try_into()?`, mapping an error if there's a mismatched type

Reply:
I understood what you meant, but I didn't like the idea of having a second error which in theory could never happen, so I've decided to use [`checked_sub`](https://doc.rust-lang.org/nightly/std/primitive.u16.html#method.checked_sub)

Reply:
Sorry I think I was unclear here, I'm OK with the subtraction (but checked sub is good!) What I really wanted to address was the `try_into().unwrap()`

Reply:
Honestly while going through `TryFrom` for converting a slice into an array is a bit ugly, handling an error that can _never_ happen looks even worse. `.try_into()` here is just [making sure](https://doc.rust-lang.org/1.77.1/src/core/array/mod.rs.html#256) that the slice returned by slicing the slice is exactly as long as the array you'd like to get. If that fails it's because the code is broken, so I don't see a graceful way if handling the error besides panicking.

I think the better solution would be to:

1. Leave it as-is
2. Add more tests to prove the panic is unreachable
3. Use [Kani](https://github.com/model-checking/kani)